lib/phil/charlie_cards/importer.ex

```diff
@@ -0,0 +1,61 @@
+defmodule Phil.CharlieCards.Importer do
+  @moduledoc """
+  Imports CharlieCard inventory into Phil.
+  """
+
+  import Ecto.Query, only: [from: 2]
+
+  alias NimbleCSV.RFC4180, as: CSV
+
+  alias Phil.CharlieCards.CharlieCard
+  alias Phil.Products.Product
+  alias Phil.Repo
+
+  @media %{"2" => :ticket, "5" => :card}
+
+  @doc """
+  Given the path of a file, this function attempts to import CharlieCards.
+
+  It returns a map of successes and failures like %{error: [], ok: []}.
+  Error entries are changesets and success entries are the inserted struct.
+  """
+  def import(path) do
+    path
+    |> File.stream!()
+    |> CSV.parse_stream()
+    |> Stream.map(&row_to_changeset/1)
+    |> Stream.chunk_every(100)
+    |> Task.async_stream(&insert_batch/1, max_concurrency: 10, ordered: false)
+    |> Enum.reduce([], fn {:ok, results}, acc -> acc ++ results end)
+    |> Enum.group_by(fn {status, _} -> status end, fn {_, value} -> value end)
+  end
+
+  defp insert_batch(changesets) do
+    Enum.map(changesets, fn changeset ->
+      if changeset.valid? do
+        Repo.insert(changeset)
+      else
+        {:error, changeset}
+      end
+    end)
+  end
+
+  defp row_to_changeset(row) do
```

Comment:
this is fine, but could also be clearer to be like:

```elixir
[
  batch_number, 
  batch_sequence_number,
  serial_number,
   medium,
   ticket_type_id,
   _,
   production_date,
   card_valid_from,
   card_valid_until,
   product_valid_from,
   product_valid_until
] = row
```

Reply:
Actually, after looking at it, this makes it more confusing (to me) because I have to destructure the list and then construct a map.

Reply:
yea it's more lines but the nice thing is it's more assertive. `Enum.at([], 8)` just returns nil, but for a CSV we don't expect there to be missing columns.

Reply:
Nil is good though because the changeset will catch it. And, I wouldn't be surprised to get nils because there is some hand manipulation involved.

Reply:
a nil value, sure, but you wont get a missing column because the CSV would be invalid

Reply:
Interestingly I just read [this](https://hexdocs.pm/nimble_csv/NimbleCSV.html#module-binary-references) in the docs which makes me wonder if we need to use `:binary.copy` on any of the strings.

Reply:
we do pass the data to another process because we spin up tasks and do it async. They are short lived so maybe it's okay but it's hard to know. I also don't know if creating a struct does copy the data in this instance I should probably look that up

Reply:
Everything is a copy in Elixir.